src/python/pants/core/util_rules/environments.py

```diff
@@ -363,5 +319,130 @@ def extract_process_config_from_environment(tgt: EnvironmentTarget) -> ProcessCo
     return ProcessConfigFromEnvironment(docker_image=docker_image)
 
 
+logger = logging.getLogger(__name__)
+
+
+class EnvironmentSensitiveOptionFieldMixin(ABC):
+    subsystem: type[Subsystem]
+    option_name: str
+
+
+_rules_for_subsystems: dict[type[Subsystem], set[UnionRule]] = {}
+
+
+def add_option_fields_for(subsystem: type[Subsystem]) -> Iterable[UnionRule]:
+    """Register environment fields for the `environment_sensitive` options of `subsystem`
+
+    This should be called in the `rules()` method in the file where `subsystem` is defined. It will
+    register the relevant fields under the `local_environment` and `docker_environment` targets.
+    """
+    if subsystem in _rules_for_subsystems:
+        return _rules_for_subsystems[subsystem]
+
+    field_rules: set[UnionRule] = set()
+
+    for option_attrname, option in _collect_options_info_extended(subsystem):
+        if option.flag_options["environment_sensitive"]:
+            field_rules.update(_add_option_field_for(subsystem, option, option_attrname))
+
+    _rules_for_subsystems[subsystem] = field_rules
+    return field_rules
+
+
+def _add_option_field_for(
+    subsystem_t: type[Subsystem], option: OptionsInfo, attrname: str
+) -> Iterable[UnionRule]:
+    option_type: type = option.flag_options["type"]
+    scope = getattr(subsystem_t, "options_scope")
+
+    # If you get a `KeyError` due to a missing value in `map` or `list_map`, it means there
+    # isn't a mapping for a given option type added at the bottom of this file. Add it there
+    # (note that currently it's not immediately clear how to support `Enum` options)
+
+    ofm: _OptionFieldMap
```

Comment:
Nit: maybe fully spell this variable out. This code is pretty magical (good magic), so having to keep track of the magic + a short var name is confusing